File: typescript/euler/bin/p0072.ts
// project euler: problem 72

/*
  See https://mathproblems123.wordpress.com/2018/05/10/sum-of-the-euler-totient-function/
*/

import { isqrt } from "../lib/math.ts";
import { range } from "../lib/util.ts";

const sumPhi = (n: number, memo: Map<string, number> = new Map()): number => {
  const trunc = Math.trunc;

  const key = String(n);
  if (memo.has(key)) {
    return memo.get(key)!;
  }

  let v = trunc(n * (n + 1) / 2);
  for (const m of range(2, isqrt(n) + 1)) {
    v -= sumPhi(trunc(n / m), memo);
  }
  for (const d of range(1, trunc(n / (isqrt(n) + 1)) + 1)) {
    v -= (trunc(n / d) - trunc(n / (d + 1))) * sumPhi(d, memo);
  }
  memo.set(key, v);

  return v;
};

export const compute = (limit: number): string =>
  String(sumPhi(limit) - sumPhi(1));

export const solve = (): string => compute(1_000_000);

File: clojure/src/project_euler/sol/p0072.clj
(ns project-euler.sol.p0072
  (:require [project-euler.lib.math :as math]))

;;; Please see the following URL.
;;;   https://mathproblems123.wordpress.com/2018/05/10/sum-of-the-euler-totient-function/
(def sum-phi
  (memoize (fn [n]
             (let [v (quot (* n (inc n)) 2)
                   sum-m (apply + (for [m (range 2 (inc (math/isqrt-long n)))]
                                    (sum-phi (quot n m))))
                   sum-d (apply + (for [d (range 1 (inc (quot n (+ (math/isqrt-long n) 1))))]
                                    (* (- (quot n d) (quot n (inc d))) (sum-phi d))))]
               (- v sum-m sum-d)))))

(defn solve
  ([]
   (solve 1000000))
  ([limit]
   (- (sum-phi limit) (sum-phi 1))))
